python/ray/util/ray_executor.py

```diff
@@ -0,0 +1,277 @@
+import itertools
+import time
+from concurrent.futures import Executor
+
+import ray
+from ray.util.annotations import PublicAPI
+
+
+@PublicAPI
+class RayExecutor(Executor):
+    """`RayExecutor` is a drop-in replacement for `ProcessPoolExecutor` and
+    `ThreadPoolExecutor` from `concurrent.futures` but distributes and executes
+    the specified tasks over a Ray cluster instead of multiple processes or
+    threads.
+    """
+
+    _shutdown_lock = False
+    """A dictionary of futures and associated Ray object references generated
+    by the current instance of RayExecutor"""
+    _futures = {}
+
+    def __init__(self, **kwargs):
+        """Initialises a new RayExecutor instance which distributes tasks over
+        a Ray cluster.
+
+        Args:
+            All keyword arguments are passed to ray.init() (see
+            https://docs.ray.io/en/latest/ray-core/package-ref.html#ray-init).
+            For example, this will connect to a cluster at the specified address:
+            .. code-block:: python
+
+                RayExecutor(address='192.168.0.123:25001')
+
+            Note: excluding an address will initialise a local Ray cluster.
+
+        """
+
+        """
+        The following is necessary because `@ray.remote` is only available at runtime.
+        """
+
+        @ray.remote
+        def remote_fn(fn, *args, **kwargs):
+            return fn(*args, **kwargs)
+
+        self.__remote_fn = remote_fn
+        self.context = ray.init(ignore_reinit_error=True, **kwargs)
```

Comment:
Should we expose context as @property method?

Reply:
The initial intent was that `RayExecutor` will be used in context (the `with` python context). So once we exit the context, ray is shutdown. I don't think we need to expose `context` as `@property` method, at least for now